### PR TITLE
Using dir locations stack for language inheritance

### PR DESCRIPTION
**Pull request by ThoWagen**
With this PR the config stack build created using the `DirLocations.ini` is also used for the languages.

### CODE REVIEW

**Comment by GrothTobias** [comment]
Hello @ThoWagen,  I wanted to implement something similar and saw this PR. 
I haven't checked out this PR to try it myself, but I believe a problem I had is also present here: translation domains that only exist in a parent directory (not vufind's basic `/languages`) will not be translated. 
A possible way to fix this is to edit `\VuFind\I18n\Translator\LanguageInitializerTrait::getTextDomains()` used in `\VuFind\i18n\Translator\TranslatorFactory`.

Can you please check this?

**Comment by demiankatz** [comment]
Thanks for bringing this up, @GrothTobias! I haven't had a chance to review this closely yet myself (there's been an incredible amount of PR activity this week!) so I'll hold off on further activity until we hear what @ThoWagen thinks.

A couple of thoughts in the meantime:

1.) Is it possible that the issue you report can/should be addressed independently of the work here? Without digging into the code, I'm not sure if this is a pre-existing problem that can be triggered in other ways, or if it only happens after making the changes found here. If it's pre-existing, maybe we can solve it in a separate PR as a prerequisite to this PR, just to keep things easy to review. But if the issues are intertwined, then of course we can just incorporate the solution here.

2.) Is there an easy way to build a test to demonstrate the problem you describe -- e.g. with some new fixtures? If so, I think having a test to cover this scenario would be very helpful. If feasible, we should try to do that as part of the work to solve the problem.

Please let me know if there's anything more I can do to be of assistance, and I'll also circle back and give this a closer code review once I'm confident the code is unlikely to significantly change in the near future. :-)

**Comment by ThoWagen** [comment]
@GrothTobias Thanks for pointing that out. @demiankatz I belive this is a problem with this PR. I did not add the local dir stack in the method mentioned by Tobias and will do so.

I also found more places where a stack of LOCAL_OVERRIDE_DIR and APPLICATION_PATH is used and where it probably would make sense to add the `DirLocationTrait`.

https://github.com/vufind-org/vufind/blob/2f54d62dc9eb8c905676a197ce592f0e7681ea14/module/VuFind/src/VuFind/Config/Locator.php#L107
https://github.com/vufind-org/vufind/blob/2f54d62dc9eb8c905676a197ce592f0e7681ea14/module/VuFind/src/VuFind/XSLT/Processor.php#L56
https://github.com/vufind-org/vufind/blob/2f54d62dc9eb8c905676a197ce592f0e7681ea14/module/VuFind/src/VuFindTest/Feature/PathResolverTrait.php#L57

What do you think about that?

And I agree that adding tests is a good idea.

File: module/VuFind/src/VuFind/I18n/Translator/Loader/ExtendedIniFactory.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/I18n/Translator/Loader/ExtendedIniFactory.php]: I only recently learned that PHP supports arrow functions, but this would be a great spot for one, reducing three lines down to:

`fn ($dirLocations) => $dirLocations['dir'] . '/languages',`

Comment (demiankatz) [reply module/VuFind/src/VuFind/I18n/Translator/Loader/ExtendedIniFactory.php]: Also, would it be more clear to use singular $dirLocation since each instance is just one location?

File: module/VuFind/src/VuFind/Config/PathResolverFactory.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Config/PathResolverFactory.php]: As noted elsewhere, I think singular $dirLocation might be more clear here.

File: module/VuFind/src/VuFind/Feature/DirLocationsTrait.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Feature/DirLocationsTrait.php]: Is it possible it's worth putting this in the VuFind\Config\Feature namespace, since it has to do with reading the directory configuration?

I also wonder if it might make sense to design this a little more like the IniReaderTrait, where it uses a helper object (e.g. something like \VuFind\Config\DirLocationsStackReader) in a property and initializes it when unset, so there's less business logic in the trait and it's easier to override that logic if desired. Maybe it would even make sense to make that a service so that there's a singleton of the reader and the stack can be centrally stored to avoid repeating the calculations in multiple contexts... though I wouldn't insist on that if it causes too much disruption in terms of factories, etc. (It could also be a future enhancement if we set the stage for it by making the design flexible to begin with).

File: module/VuFind/src/VuFind/Feature/DirLocationsTrait.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Feature/DirLocationsTrait.php]: I think this should be indented further.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen -- I agree that adding this logic to all affected code makes sense for consistency. I took the time to give it a closer review; see below for a few comments.

File: module/VuFind/src/VuFind/Feature/DirLocationsTrait.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Feature/DirLocationsTrait.php]: The DirLocations stack resolving needs to be used in multiple locations where no direct access to the service locator is given (LanguageInitializerTrait,VuFind/Config/Locator,PathResolverTrait, etc.). So we would have to refactor the workflows in some way. I'm not sure if that is worth the effort. Or am I missing an easy solution here?

And I don't understand how moving the business logic into a helper object would make it easier to override it. Wouldn't you still have to create your own trait and helper object and apply it everywhere on your own anyway?

I don't reject your suggestion and also think it would be good if this part would be overridable but just don't know how to accomplish it.

**Comment by ThoWagen** [comment]
@demiankatz Regarding the tests I also don't know where to do this. As far as I know there is no complete 'Translator' test that could be extended for the case described by Tobias. A simple unittest is probably not enough. Should I just create a Mink testcase for that and use arbitrary translations? I guess that would also require creating something like "local2" and adding a DirLocations.ini to "local".

File: module/VuFind/src/VuFind/Feature/DirLocationsTrait.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/Feature/DirLocationsTrait.php]: My original suggestion was based on the fact that the trait was only being used in factories, which should always have access to the top-level container. I haven't done the searching to see how many other classes are impacted based on LanguageInitializerTrait, etc. If it's a relatively small number, it might still be feasible to give the trait a getter and setter for a helper object -- the setter could be used to inject the helper from the container in relevant factories, and the getter could instantiate a default object if no setter was called. If a really large number of files are involved, it might be possible to create an "AwareInterface" and an initializer class to auto-inject the dependency, though I've been trying to move away from that pattern, as it's a bit too "magical" for my liking.

But in any case, the reason moving the business logic into a helper object would make things easier to override is that you can override a helper object in the service locator, but you can't override a trait... so if all the trait is doing is managing dependency injection and wrapping around a helper object, it should be possible to replace the helper without needing to change the trait everywhere it's used (as long as factories or auto-injection are set up correctly).

Is that any help, or am I missing/misunderstanding part of your concern?

**Comment by demiankatz** [comment]
> @demiankatz Regarding the tests I also don't know where to do this. As far as I know there is no complete 'Translator' test that could be extended for the case described by Tobias. A simple unittest is probably not enough. Should I just create a Mink testcase for that and use arbitrary translations? I guess that would also require creating something like "local2" and adding a DirLocations.ini to "local".

That might be a possibility. Since the parent directory can be anywhere, I would suggest creating the parent directory in one of the test fixture directories -- then I think you could just create a new helper method in the base Mink test case to create (and auto-remove) DirLocations.ini as part of a test, similar to the existing config handling.

File: module/VuFind/src/VuFind/Feature/DirLocationsTrait.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Feature/DirLocationsTrait.php]: The LanguageInitializerTrait is only used in two places. My concerns were also the other places where the stack is resolved. However, the VuFind/Config/Locator is deprecated and replaced by the PathResolver and PathResolverTrait is only used in tests. So we probably don't need the logic there. I don't know what XSLT/Processor.php is used for. We could also ignore that.

What do you think about adding a "getLocalConfigDirStack" getter to the PathResolver and just using that instead of a new service?

Comment (demiankatz) [reply module/VuFind/src/VuFind/Feature/DirLocationsTrait.php]: Yes, it should be safe not to worry about deprecated logic.

The XSLT/Processor is, I believe, used by the XML import (and also possibly the MARC staff view).

Adding a new method to the PathResolver seems reasonable to me!

File: module/VuFind/src/VuFind/I18n/Translator/Loader/ExtendedIniFactory.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/I18n/Translator/Loader/ExtendedIniFactory.php]: Good Idea! Done.

File: module/VuFind/src/VuFind/Config/PathResolverFactory.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Config/PathResolverFactory.php]: Due to other changes this became obsolete.

File: module/VuFind/src/VuFind/Feature/DirLocationsTrait.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Feature/DirLocationsTrait.php]: I implemented this now as suggested by adding a "getLocalConfigDirStack" method to the PathResolver.

File: module/VuFind/src/VuFind/Feature/DirLocationsTrait.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Feature/DirLocationsTrait.php]: Fixed

**Comment by ThoWagen** [comment]
@demiankatz I believe I implemented all the suggestions including some tests and fixing the problem with the translations domains.

File: module/VuFind/src/VuFind/I18n/Translator/LanguageInitializerTrait.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/I18n/Translator/LanguageInitializerTrait.php]: Since we're not setting this via constructor, we ought to make it optional:

```suggestion
     * @var ?PathResolver
     */
    protected ?PathResolver $pathResolver = null;
```

File: module/VuFind/src/VuFind/I18n/Translator/LanguageInitializerTrait.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/I18n/Translator/LanguageInitializerTrait.php]: We should probably account for the possibility of `$this->pathResolver` being unset. Maybe in that case we should use an empty array and issue a warning via `error_log` or some other mechanism.

File: module/VuFind/src/VuFindTest/Integration/MinkTestCase.php
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Integration/MinkTestCase.php]: You don't need to indent and wrap this:

```suggestion
     * (as opposed to extending it with new settings)?
```

File: module/VuFind/src/VuFindTest/Integration/MinkTestCase.php
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Integration/MinkTestCase.php]: If we're adding types, might as well do the return value too.

```suggestion
    protected function changeConfigFile(string $configName, array $settings, bool $replace = false): void
```

File: module/VuFind/src/VuFindTest/Integration/MinkTestCase.php
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Integration/MinkTestCase.php]: Adding return value here too.

```suggestion
    protected function writeConfigFile(string $path, array $settings, bool $replace = false): void
```

File: module/VuFindConsole/src/VuFindConsole/Command/ScheduledSearch/NotifyCommandFactory.php
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/ScheduledSearch/NotifyCommandFactory.php:92]: I see changes to the factory and the test but not to NotifyCommand itself; are these changes still relevant/needed? Am I missing something?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen, this is looking very good! See below for a few minor suggestions and questions.

File: module/VuFindConsole/src/VuFindConsole/Command/ScheduledSearch/NotifyCommandFactory.php
Comment (demiankatz) [reply module/VuFindConsole/src/VuFindConsole/Command/ScheduledSearch/NotifyCommandFactory.php:92]: Never mind, I see that this is related to LanguageInitializerTrait; makes sense to me now! :-)

File: module/VuFind/src/VuFind/I18n/Translator/LanguageInitializerTrait.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/I18n/Translator/LanguageInitializerTrait.php]: Yes, you are right. That's fixed using `error_log` now.

**Comment by demiankatz** [comment]
Thanks, @ThoWagen, this looks good to me now!

@GrothTobias, do you have any further comments/suggestions before this gets merged?

**Comment by GrothTobias** [comment]
I think this looks good. 
I just wanted to point out a possible problem since it occured to me just before I saw this PR.

**Review by demiankatz — APPROVED** [review]

**Pull request merged**
